crates/sui-core/src/transaction_manager.rs

```diff
@@ -94,7 +96,121 @@ impl LockQueue {
     }
 }
 
-#[derive(Default)]
+struct CacheInner {
+    versioned_cache: LruCache<ObjectID, SequenceNumber>,
+    // we cache packages separately, because they are more expensive to look up in the db, so we
+    // don't want to evict packages in favor of mutable objects.
+    unversioned_cache: LruCache<ObjectID, ()>,
+
+    max_size: usize,
+    metrics: Arc<AuthorityMetrics>,
+}
+
+impl CacheInner {
+    fn new(max_size: usize, metrics: Arc<AuthorityMetrics>) -> Self {
+        Self {
+            versioned_cache: LruCache::unbounded(),
+            unversioned_cache: LruCache::unbounded(),
+            max_size,
+            metrics,
+        }
+    }
+}
+
+impl CacheInner {
+    fn shrink(&mut self) {
+        while self.versioned_cache.len() > self.max_size {
+            self.versioned_cache.pop_lru();
+            self.metrics
+                .transaction_manager_object_cache_evictions
+                .inc();
+        }
+        while self.unversioned_cache.len() > self.max_size {
+            self.unversioned_cache.pop_lru();
+            self.metrics
+                .transaction_manager_object_cache_evictions
+                .inc();
+        }
+    }
+
+    fn insert(&mut self, object: &InputKey) {
+        if let Some(version) = object.1 {
+            if let Some((previous_id, previous_version)) =
+                self.versioned_cache.push(object.0, version)
+            {
+                if previous_id == object.0 && previous_version > version {
+                    // do not allow highest known version to decrease
+                    self.versioned_cache.put(object.0, previous_version);
+                } else {
+                    self.metrics
+                        .transaction_manager_object_cache_evictions
+                        .inc();
+                }
+            }
+        } else if let Some((previous_id, _)) = self.unversioned_cache.push(object.0, ()) {
+            if previous_id != object.0 {
```

Comment:
since the value in this cache is `()` i expect that it would never happen, but I'm not entirely sure. If it can happen, it would not make sense to count it as an eviction since it would be getting replaced with the same value.

Reply:
actually, i checked the code and it can return `()` if you insert `()`, it never checks if the values are equal

Reply:
Sounds good. Maybe adding a comment on not counting setting the same value as eviction would be useful but it is minor.

Reply:
But in general, I thought about if it is possible to re-insert the same package or lower versions objects, and it seems it is impossible if cache is only inserted on `objects_available()` and possible otherwise.